### ISSUE TITLE
Default `class` `@Parameter`s allow information sharing.

### ISSUE THREAD

**Issue by theMomax** [Apodini/Apodini#122]
**Describe the bug**
All `Handler`s use the same instance of a `@Parameter`'s default value. This is no problem for value-types. The `@Parameter`'s `set`ter is not available to the user, thus value-types cannot be altered and sharing the default value is fine.

However, reference-types (i.e. `class` objects) can be altered nevertheless. So, when using a `class` type on an `@Parameter` with a default value, `Handler`s share this default value **and** can alter it.

**To Reproduce**

Add the following handler to your code at ["v1", "greet"]:
```swift
class ClassString: Codable {
    var name: String
    
    init(name: String) {
        self.name = name
    }
}

struct Greeter: Handler {
        @Parameter var name: ClassString = ClassString(name: "Apodini")
        @Parameter var overwrite: Bool = false

        func handle() -> String {
            if overwrite {
                self.name.name = "Max"
            }
            return "Hi, \(name.name)!"
        }
    }
```
Open a WebSocket connection:
`websocat ws://127.0.0.1:8080/apodini/websocket`
Send the following messages:
```json
{"context":"a21f0915-825a-422e-b8dc-1876234cd08d","endpoint":"v1.greet"}
{"context":"a21f0915-825a-422e-b8dc-1876234cd08d","parameters": {"overwrite": true}}
```
It should answer with
```json
{"context":"A21F0915-825A-422E-B8DC-1876234CD08D","content":"Hi, Max!"}
{"context":"A21F0915-825A-422E-B8DC-1876234CD08D"}
```

Open a second WebSocket connection:
`websocat ws://127.0.0.1:8080/apodini/websocket`
Send the following messages (not that overwrite is `false`):
```json
{"context":"a21f0915-825a-422e-b8dc-1876234cd08e","endpoint":"v1.greet"}
{"context":"a21f0915-825a-422e-b8dc-1876234cd08e","parameters": {}}
```
It should answer with the same content.
```json
{"context":"A21F0915-825A-422E-B8DC-1876234CD08E","content":"Hi, Max!"}
{"context":"A21F0915-825A-422E-B8DC-1876234CD08E"}
```
Note that this behavior is not specific to the WebSocket exporter or its concept of context. 

**Expected behavior**
As the user of Apodini I imagine `Handler`s to be instantiated for each connection and thus `@Parameter`s to never be shared.

I see several ways to deal with this:

##### It's OK
We just say that is ok as no developer should alter `@Parameter`s and we just cannot prevent that for classes.

##### Class Not Supported
We add a precondition to `@Parameter` checking on startup that the contained type is no reference-type. Could we check this recursively so that e.g. `Optional<SomeClassType>` would also be detected?

##### Auto-Closures
My first idea was to use an auto-closure on the `@Parameter`'s `defaultValue` initializers, so we are able to create a new instance each time a connection is created. This would not prevent the developer from using a singleton (`@Parameter var name: SomeClassType = SomeService.shared`), but then it is clear to everyone reading it, that this value will be shared across `Handler` instances.

The problem with this approach is that it doesn't compile: I'm getting a Segmentation Fault 11

My code is on [experimental/autoclosure-initializers-for-parameter](https://github.com/Apodini/Apodini/tree/experimental/autoclosure-initializers-for-parameter).

**Screenshots**
![image](https://user-images.githubusercontent.com/21169289/104303202-1527f800-54ca-11eb-8443-c23a29d60aae.png)

**Environment (enter your versions)**

 - Swift Version: 5.3.2 (swiftlang-1200.0.45 clang-1200.0.32.28)
 - Xcode Version: 12.3 (12C33)

**Additional context**
I came across this issue when thinking about implementations for `@State`, where this problem becomes rather critical, since altering a default value would be pretty much the most common operation there.

**Comment by bauer-andreas** [comment]
Just quickly glancing over it. I saw that you call the closure when construction the `EndpointParameter`, which creates the same problem, as shared instances would be used?
We should probably keep the closure in the `EndpointParameter` and call it for every request handled (or when its used e.g. in the OpenAPI exporter to describe the Endpoint).

**Comment by theMomax** [comment]
> Just quickly glancing over it. I saw that you call the closure when construction the `EndpointParameter`, which creates the same problem, as shared instances would be used?
> We should probably keep the closure in the `EndpointParameter` and call it for every request handled (or when its used e.g. in the OpenAPI exporter to describe the Endpoint).

Yes, that is just temporary as I didn't want to add more changes before I get it to compile.

**Comment by theMomax** [comment]
We agreed that the **Auto-closure** approach is the most desirable. It is implemented in #125.
